lua/entities/gmod_wire_expression2/core/timer.lua

```diff
@@ -1,121 +1,142 @@
-	for name,_ in pairs(self.data['timer'].timers) do
-		RemoveTimer(self, name)
+	for name in pairs(Timers[self]) do
+		removeTimer(self, name)
 	end
-end)
 
-/******************************************************************************/
+	Timers[self] = nil
+end)
 
 __e2setcost(20)
 
+---@param self RuntimeContext
+local function MAKE_TRIGGER(id, self)
+	return function()
+		self.data.timer = id
+		removeTimer(self, id)
+
+		if self.entity and self.entity.Execute then
+			self.entity:Execute()
+		end
+
+		self.data.timer = nil
+	end
+end
+
+[deprecated = "Use the timer function with callbacks instead"]
 e2function void interval(rv1)
-	AddTimer(self, "interval", rv1)
+	addTimer(self, "interval", rv1 / 1000, 1, MAKE_TRIGGER("interval", self))
 end
 
+[deprecated = "Use the timer function with callbacks instead"]
 e2function void timer(string rv1, rv2)
-	AddTimer(self, rv1, rv2)
+	addTimer(self, rv1, rv2 / 1000, 1, MAKE_TRIGGER(rv1, self))
 end
 
 __e2setcost(5)
 
 e2function void stoptimer(string rv1)
-	RemoveTimer(self, rv1)
+	removeTimer(self, rv1)
 end
 
 __e2setcost(1)
 
-[nodiscard]
+[nodiscard, deprecated = "Use the timer function with callbacks instead"]
 e2function number clk()
-	return self.data.timer.runner == "interval" and 1 or 0
+	return self.data.timer == "interval" and 1 or 0
 end
 
-[nodiscard]
+[nodiscard, deprecated = "Use the timer function with callbacks instead"]
 e2function number clk(string rv1)
-	return self.data.timer.runner == rv1 and 1 or 0
+	return self.data.timer == rv1 and 1 or 0
 end
 
-[nodiscard]
+[nodiscard, deprecated = "Use the timer function with callbacks instead"]
 e2function string clkName()
-	return self.data.timer.runner or ""
+	return self.data.timer or ""
 end
 
+__e2setcost(5)
+
+[deprecated = "Use the timer function with callbacks instead"]
 e2function array getTimers()
-	local ret = {}
+	local ret, timers = {}, Timers[self]
+	if not timers then return ret end
+
+	self.prf = self.prf + #timers * 2
+
 	local i = 0
-	for name in pairs( self.data.timer.timers ) do
+	for name in pairs(timers) do
 		i = i + 1
 		ret[i] = name
 	end
-	self.prf = self.prf + i * 5
+
 	return ret
 end
 
 e2function void stopAllTimers()
-	for name in pairs(self.data.timer.timers) do
-		self.prf = self.prf + 5
-		RemoveTimer(self,name)
+	local timers = Timers[self]
+	if not timers then return end
+
+	self.prf = self.prf + #timers * 2
+
+	for name in pairs(timers) do
+		removeTimer(self, name)
 	end
 end
 
-/******************************************************************************/
+--[[
+	Timers 2.0
+]]
+
+__e2setcost(10)
+
+-- Create "anonymous" timer using address of arguments, which should be different for each function call.
+-- Definitely hacky, but should work properly. I think this is better than just incrementing a number infinitely.
+e2function void timer(number delay, function callback)
+	local fn, ent = callback:Unwrap("", self), self.entity
+	addTimer(self,("%p"):format(args), delay, 1, function()
```

Comment:
I'd recommend a more unique algo. In my testing I noticed there's maybe a 2% chance of dropping a timer on the same execution and that will just make a lot of headache in the future. Not to mention I believe this means you could accidentally override a timer anyway if the address is reused.

Reply:
> I'd recommend a more unique name algo. In my testing I noticed there's maybe a 2% chance of dropping a timer on the same execution and that will just make a lot of headache in the future. Not to mention I believe this means you could accidentally override a timer anyway if the address is reused.

Already been done. Not commited. I wouldn't recommend reviewing this (or any draft prs for that matter)

Reply:
I am impatient and therefore playing with them.